lib/sprockets/utils.rb

```diff
@@ -100,18 +98,18 @@ def string_end_with_semicolon?(str)
     #
     # Returns buf String.
     def concat_javascript_sources(buf, source)
-      if source.bytesize > 0
-        buf << source
+      return buf if source.bytesize < 0
 
-        # If the source contains non-ASCII characters, indexing on it becomes O(N).
-        # This will lead to O(N^2) performance in string_end_with_semicolon?, so we should use 32 bit encoding to make sure indexing stays O(1)
-        source = source.encode(Encoding::UTF_32LE) unless source.ascii_only?
+      buf << source
+      # If the source contains non-ASCII characters, indexing on it becomes O(N).
+      # This will lead to O(N^2) performance in string_end_with_semicolon?, so we should use 32 bit encoding to make sure indexing stays O(1)
+      source = source.encode(Encoding::UTF_32LE) unless source.ascii_only?
+      return buf if string_end_with_semicolon?(source)
 
-        if !string_end_with_semicolon?(source)
-          buf << ";\n"
-        elsif source[source.size - 1].ord != 0x0A
-          buf << "\n"
-        end
+      if whitespace = WHITESPACE_ORDINALS[buf[-1].ord]
+        buf[-1] = ";#{whitespace}"
+      else
+        buf << ";"
       end
```

Comment:
This is officially a "there be dragons" method. Your PR commentary would be appropriate for discussion in code comments here, too, so it's clear what conditions are at play and where they're implemented.

Switching from conditional expressions to conditional control flow makes this simpler to read in imperative style, but it's harder (for me) to parse and understand.